subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionPlan.java

```diff
@@ -661,12 +662,16 @@ private ResourceLock getOrCreateProjectLock(Project project) {
     private boolean canRunWithCurrentlyExecutedTasks(TaskInfo taskInfo) {
         Set<String> candidateTaskDestroyables = getDestroyablePaths(taskInfo);
 
-        if (!candidateTaskDestroyables.isEmpty() && taskInfo.hasOutputs()) {
-            throw new IllegalStateException("Task " + taskInfo.getTask().getIdentityPath() + " has both outputs and destroyables defined.  A task can define either outputs or destroyables, but not both.");
-        }
-
-        if (!candidateTaskDestroyables.isEmpty() && taskInfo.hasFileInputs()) {
-            throw new IllegalStateException("Task " + taskInfo.getTask().getIdentityPath() + " has both inputs and destroyables defined.  A task can define either inputs or destroyables, but not both.");
+        if (!candidateTaskDestroyables.isEmpty()) {
+            if (taskInfo.hasOutputs()) {
+                throw new IllegalStateException("Task " + taskInfo.getTask().getIdentityPath() + " has both outputs and destroyables defined.  A task can define either outputs or destroyables, but not both.");
```

Comment:
Should we define a method for the exception message, too? Something like
```java
String conflictingWithDestroyableMessage(Task task, String conflictingPropertyType) {
     return "Task " + taskInfo.getTask().getIdentityPath() + " has both " + conflictingPropertyType + " and destroyables defined.  A task can define either " + conflictingPropertyType + " or destroyables, but not both.";
}
```

Reply:
Let's do that the next time we change this code.